libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -127,6 +119,18 @@ string IRGenerator::generate(Block const& _block)
 	return generator.code();
 }
 
+void IRGenerator::collectFunctions(ContractDefinition const& _contract)
+{
+	for (auto const* contract: _contract.annotation().linearizedBaseContracts)
+		for (auto const* fun: contract->definedFunctions())
+			if (fun->isVisibleViaContractTypeAccess())
```

Comment:
Why only these?

Reply:
Actually I think we do not need this part at all. The first functions should be inserted by
 - `constructorCode` for the creation context and
 - `dispatchRoutine` for the runtime context

Reply:
I start by collecting only the functions that must be present even if they're not called from other functions. All the others will get added when calls to them are detected in ``IRGeneratorForStatements::endVisit()``.

At first I assumed that I only need external functions but this does not include `constructor`, `fallback` or `receive`. They're called after the dispatch routine but AFAIK that routine does not trigger `endVisit()` so I need to include them here. So `isVisibleViaContractTypeAccess()` seems to be what I need.

Actually, now that I think about it, maybe I should modify `IRGenerator::dispatchRoutine()` to generate code for them and `isPartOfExternalInterface()` here instead?

Also, please take a look at my remark about internal constructors here: https://github.com/ethereum/solidity/pull/8663#discussion_r408151089

Reply:
@chriseth ok, seems like a better way to do it. I'll change it.

Reply:
Yeah, I think `dispatchRoutine` should not directly call `generateFunction` but instead just add the function to the queue. Since the dispatch routine iterates over all external interface funcitons, the loop in this file here is not required. The same is true for the fallback and receive functions and the constructors.